otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py

```diff
@@ -0,0 +1,157 @@
+"""DetCon loss."""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+# pylint: disable=no-name-in-module, not-callable
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
+from mmseg.models.builder import LOSSES
+from torch import nn
+
+
+def manual_cross_entropy(logits, labels, weight):
+    """Manually calculate weighted cross entropy."""
+    cross_entropy = -weight * torch.sum(labels * F.log_softmax(logits, dim=-1), dim=-1)
+    return torch.mean(cross_entropy)
+
+
+@LOSSES.register_module
+class DetConLoss(nn.Module):
+    """Modified from https://github.com/deepmind/detcon/blob/main/utils/losses.py.
+
+    Compute the NCE scores from pairs of predictions and targets.
+    This implements the batched form of the loss described in
+    Section 3.1, Equation 3 in https://arxiv.org/pdf/2103.10957.pdf.
+
+    Args:
+        temperature: (float) the temperature to use for the NCE loss.
+        use_replicator_loss (bool): use cross-replica samples.
+    """
+
+    def __init__(
+        self, temperature: float = 0.1, use_replicator_loss: bool = True, ignore_index: int = 255
+    ):  # pylint: disable=unused-argument
+        super().__init__()
+        assert temperature > 0
+        self.temperature = torch.tensor(temperature)
+        self.use_replicator_loss = use_replicator_loss
+
+    # pylint: disable=too-many-statements, too-many-locals, too-many-arguments
+    def forward(self, pred1, pred2, target1, target2, pind1, pind2, tind1, tind2, local_negatives=True):
+        """Forward loss.
+
+        Args:
+            pred1 (Tensor): (b, num_samples, d) the prediction from first view.
+            pred2 (Tensor): (b, num_samples, d) the prediction from second view.
+            target1 (Tensor): (b, num_samples, d) the projection from first view.
+            target2 (Tensor): (b, num_samples, d) the projection from second view.
+            pind1 (Tensor): (b, num_samples) mask indices for first view's prediction.
+            pind2 (Tensor): (b, num_samples) mask indices for second view's prediction.
+            tind1 (Tensor): (b, num_samples) mask indices for first view's projection.
+            tind2 (Tensor): (b, num_samples) mask indices for second view's projection.
+            local_negatives (bool): whether to include local negatives.
```

Comment:
I believe there must be some smarter ways to express them.
To do this, we can do below as 
for i, (v0, v1) in enumerate(list(itertools.product(view0, view1))):
    same_obj[i] = make_same_obj(v0, v1)

Reply:
I agree there are so many parts to be refactored.
First of all, I changed the official code from jax to pytorch and left it as it is.
I'm not sure which way is more efficient and intuitive.
But there are some parts to be refactored, such as pipelines (currently: np -> PIL -> np, it is required to be changed.) and modules (for instance, `SelfSLMLP` is used for both cls and seg, and it is necessary to combine them.), so DetCon loss function simplification can be processed with the above process together.

Reply:
Done.